pil-analyzer/src/traits_processor.rs

```diff
@@ -0,0 +1,254 @@
+use core::panic;
+use std::collections::HashMap;
+
+use itertools::Itertools;
+use powdr_ast::{
+    analyzed::{
+        Expression, FunctionValueDefinition, Identity, PolynomialReference, Reference, Symbol,
+        TypedExpression,
+    },
+    parsed::{types::Type, visitor::AllChildren, SelectedExpressions, TraitImplementation},
+};
+
+pub struct TraitsProcessor<'a> {
+    definitions: &'a mut HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+    identities: &'a mut Vec<Identity<SelectedExpressions<Expression>>>,
+    implementations: &'a HashMap<String, Vec<TraitImplementation<Expression>>>,
+}
+
+impl<'a> TraitsProcessor<'a> {
+    pub fn new(
+        definitions: &'a mut HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+        identities: &'a mut Vec<Identity<SelectedExpressions<Expression>>>,
+        implementations: &'a HashMap<String, Vec<TraitImplementation<Expression>>>,
+    ) -> Self {
+        Self {
+            definitions,
+            identities,
+            implementations,
+        }
+    }
+
+    pub fn traits_resolution(&mut self) {
+        let keys: Vec<String> = self
+            .definitions
+            .iter()
+            .filter(|(_, (_, def))| {
+                !matches!(
+                    def,
+                    None | Some(FunctionValueDefinition::TraitFunction(_, _))
+                )
+            })
+            .map(|(name, _)| name.clone())
+            .collect();
+
+        for name in keys {
+            self.resolve_trait(&name);
+        }
+
+        self.resolve_trait_identities();
+    }
+
+    fn resolve_trait_identities(&mut self) {
+        let mut updates = Vec::new();
+        for (index, identity) in self.identities.iter().enumerate() {
+            let refs_in_identity = self.collect_refs_in_identity(identity);
+            for collected_ref in refs_in_identity {
+                let (trait_name, resolved_impl_pos) =
+                    self.resolve_trait_function(collected_ref.clone());
+                updates.push((index, trait_name, resolved_impl_pos));
+            }
+        }
+
+        for (index, trait_name, resolved_impl_pos) in updates {
+            let identity = &mut self.identities[index];
+            TraitsProcessor::update_references_in_identity(
+                identity,
+                &trait_name,
+                &resolved_impl_pos,
+            );
+        }
+    }
+
+    fn collect_refs_in_identity(
```

Comment:
We do something similar for the type checker in pil_analyzer.rs - maybe we can share the routine and change the interface of the trait processor to just get a vector of expressions to process?